### PR TITLE
[youporn] Fix extraction of metadata

### LINKED ISSUE #2701: [YouPorn] Unable to download JSON metadata: HTTP Error 403: Forbidden
### Checklist

- [X] I'm reporting a broken site
- [X] I've verified that I'm running yt-dlp version **2022.02.04**. ([update instructions](https://github.com/yt-dlp/yt-dlp#update))
- [X] I've checked that all provided URLs are alive and playable in a browser
- [X] I've checked that all URLs and arguments with special characters are [properly quoted or escaped](https://github.com/ytdl-org/youtube-dl#video-url-contains-an-ampersand-and-im-getting-some-strange-output-1-2839-or-v-is-not-recognized-as-an-internal-or-external-command)
- [X] I've searched the [bugtracker](https://github.com/yt-dlp/yt-dlp/issues?q=) for similar issues including closed ones. DO NOT post duplicates
- [X] I've read the [guidelines for opening an issue](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#opening-an-issue)
- [X] I've read about [sharing account credentials](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#are-you-willing-to-share-account-details-if-needed) and I'm willing to share it if required

### Region

_No response_

### Description

Error happens using whatever YP link and download stops.
With `Youtube-dl` a different error is reported: `WARNING: unable to extract upload date; please report this issue...` but the download starts successfully.

### Verbose log

```shell
C:\temp\YT-DLP>yt-dlp -vU "https://www.youporn.com/watch/16516654/sego-lamico-di-mio-marito-al-ristorante-e-non-si-accorge-di-niente/"
[debug] Command-line config: ['-vU', 'https://www.youporn.com/watch/16516654/sego-lamico-di-mio-marito-al-ristorante-e-non-si-accorge-di-niente/']
[debug] Encodings: locale cp1252, fs utf-8, out utf-8, err utf-8, pref cp1252
[debug] yt-dlp version 2022.02.04 [c1653e9] (win_exe)
[debug] Python version 3.8.10 (CPython 64bit) - Windows-10-10.0.19044-SP0
[debug] exe versions: ffmpeg 4.2.3, ffprobe 4.2.3
[debug] Optional libraries: Cryptodome, mutagen, sqlite, websockets
[debug] Proxy map: {}
Latest version: 2022.02.04, Current version: 2022.02.04
yt-dlp is up to date (2022.02.04)
[debug] [YouPorn] Extracting URL: https://www.youporn.com/watch/16516654/sego-lamico-di-mio-marito-al-ristorante-e-non-si-accorge-di-niente/
[YouPorn] sego-lamico-di-mio-marito-al-ristorante-e-non-si-accorge-di-niente: Downloading JSON metadata
ERROR: [YouPorn] 16516654: Unable to download JSON metadata: HTTP Error 403: Forbidden (caused by <HTTPError 403: 'Forbidden'>); please report this issue on  https://github.com/yt-dlp/yt-dlp , filling out the "Broken site" issue template properly. Confirm you are on the latest version using -U (caused by <HTTPError 403: 'Forbidden'>); please report this issue on  https://github.com/yt-dlp/yt-dlp , filling out the "Broken site" issue template properly. Confirm you are on the latest version using -U
  File "yt_dlp\extractor\common.py", line 730, in _request_webpage
  File "yt_dlp\YoutubeDL.py", line 3558, in urlopen
  File "urllib\request.py", line 531, in open
  File "urllib\request.py", line 640, in http_response
  File "urllib\request.py", line 569, in error
  File "urllib\request.py", line 502, in _call_chain
  File "urllib\request.py", line 649, in http_error_default


C:\temp\YT-DLP>
```

### PR DESCRIPTION
**Pull request by marieell**
### Before submitting a *pull request* make sure you have:
- [x] At least skimmed through [contributing guidelines](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#developer-instructions) including [yt-dlp coding conventions](https://github.com/yt-dlp/yt-dlp/blob/master/CONTRIBUTING.md#yt-dlp-coding-conventions)
- [x] [Searched](https://github.com/yt-dlp/yt-dlp/search?q=is%3Apr&type=Issues) the bugtracker for similar pull requests
- [x] Checked the code with [flake8](https://pypi.python.org/pypi/flake8)

### In order to be accepted and merged into yt-dlp each piece of code must be in public domain or released under [Unlicense](http://unlicense.org/). Check one of the following options:
- [x] I am the original author of this code and I am willing to release it under [Unlicense](http://unlicense.org/)
- [ ] I am not the original author of this code but it is in public domain or released under [Unlicense](http://unlicense.org/) (provide reliable evidence)

### What is the purpose of your *pull request*?
- [x] Bug fix
- [ ] Improvement
- [ ] New extractor
- [ ] New feature

---

Closes #2701. Youporn sends a JS challenge with TLS1.3-

### CODE REVIEW

File: yt_dlp/extractor/youporn.py [hunk 1]
```diff
@@ -114,43 +114,51 @@ def _real_extract(self, url):
...
-        view_count = None
-        views = self._search_regex(
-            r'(<div[^>]+\bclass=["\']js_videoInfoViews["\']>)', webpage,
-            'views', default=None)
-        if views:
-            view_count = str_to_int(extract_attributes(views).get('data-value'))
-        comment_count = str_to_int(self._search_regex(
+            headers={'Cookie': 'age_verified=1', 'Ytdl-old-tls': True})
+
+        data = self._json_ld(self._search_regex(r'(?s)<script type="application/ld\+json">(.*?)</script>', webpage, 'JSON LD', fatal=False), video_id, fatal=False) or {}
```
Comment (dirkf) [inline yt_dlp/extractor/youporn.py:119]: Use `self._search_json_ld(webpage, video_id, expected_type='Video', fatal=False)` ?

**Comment by pukkandan** [comment]
1. The idea of creating a new context is interesting, but more consideration must be given for how it is implemented. (Related: #258, #778, #2043). For the time-being, implement `PROTOCOL_TLSv1_2` into `--legacy-server-connect` instead. You can catch the relevant error in the extractor and ask user to apply this option.
2. Why use `context.post_handshake_auth`? afaik, it is not applicable for us
3. Use `merge_dicts` instead of all `if not data[...]: data[...] = ...` in extractor code. The same can be used for the final `update`

**Comment by coletdjnz** [comment]
> 1. The idea of creating a new context is interesting, but more consideration must be given for how it is implemented. (Related: [[Feature Request]Ability to use a separate file which stores names of cookie files to use and either extractor/domain #258](https://github.com/yt-dlp/yt-dlp/issues/258), [[options] Add --legacy-server-connect #778](https://github.com/yt-dlp/yt-dlp/pull/778), [[ceskatelevize.cz] Cannot download manifest - SSLV3_ALERT_HANDSHAKE_FAILURE #2043](https://github.com/yt-dlp/yt-dlp/issues/2043)). For the time-being, implement `PROTOCOL_TLSv1_2` into `--legacy-server-connect` instead. You can catch the relevant error in the extractor and ask user to apply this option.

I'll have a look into the creating a new context thing. Along with `--legacy-server-connect`, maybe we can add a header similar to what OP did that force enables it for the request. Will need to take into consideration #2861 and #3668 too. 

> Why use context.post_handshake_auth? afaik, it is not applicable for us

yeah this looks to be related to the client cert auth which wasn't yet implemented when this PR was made. that said, I prob need to look into that too. it's implemented by default in urllib3 https://github.com/urllib3/urllib3/commit/6a626be4ff623c25270e20db9002705bf4504e4e

**Comment by pukkandan** [comment]
https://github.com/yt-dlp/yt-dlp/issues/2701 has been fixed in a different way, so only the site enhancements is remaining from this PR

**Comment by marieell** [comment]
> 3\. Use `merge_dicts` instead of all `if not data[...]: data[...] = ...` in extractor code. The same can be used for the final `update`

But that would always execute all these extractions, even if they are not necessary?

**Comment by pukkandan** [comment]
Is that an issue?

**Comment by marieell** [comment]
It's a lot of possibly unneeded string matching for in my opinion not much of a style improvement. Well. What do you want it to look like?
```python
data = merge_dicts(data, { 'title': self._html_search_regex(
// …
)})
```

**Comment by pukkandan** [comment]
Ive pushed some changes. pls add a test

**Comment by pukkandan** [comment]
Actually, I just realized that the whole PR is really just one line. Everything else is unnecessary changes!

**Comment by marieell** [comment]
Alright, I squashed the commits.

**Comment by marieell** [comment]
> Pls avoid force pushing in future

Is that documented somewhere? Most projects I know strongly encourage cleaning up commits in PRs.

File: yt_dlp/extractor/youporn.py [hunk 2]
```diff
@@ -64,6 +65,25 @@ class YouPornIE(InfoExtractor):
...
+            'duration': 298.0,
+            'ext': 'mp4',
+            'upload_date': '20201123',
+            'uploader': 'Ersties',
+            'tags': [],
+            'thumbnail': 'https://fi1.ypncdn.com/202011/23/16290308/original/8/tinderspecial-trailer1-8(m=eaAaaEPbaaaa).jpg',
+            'timestamp': 1606089600,
+            'title': 'Tinder In Real Life',
+            'view_count': int,
+
```
Comment (pukkandan) [inline yt_dlp/extractor/youporn.py:85]:
```suggestion
```

**Review by pukkandan — APPROVED** [review]

**Comment by pukkandan** [comment]
> Is that documented somewhere?

No

> Most projects I know strongly encourage cleaning up commits in PRs.

It makes reviewing harder. And we squash and merge anyway

**Pull request merged**